packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to either define the field to sort by, or provide a custom sort routine. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```
+<nimble-table>
+    <nimble-table-column-text valueKey="name"></nimble-table-column-text>
```

Comment:
The difference that I see is that the key being referred to here is for a particular value within the table. Additionally, we expect to have some column definitions supply multiple keys, and using something like `data-key1` and `data-key2` doesn't seem like the pattern we want to promote. Instead, I think the pattern is that the individual column APIs provide contextual names. For the "text" column, I'm not sure there's a big difference between "data" and "value", but perhaps `text-key` is better? Not sure I want to propose the right API name for the icon column since we're defining that elsewhere.